src/Components/Components/src/Virtualization/InfiniteScroll.cs

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Components.Virtualization
+{
+    /// <summary>
+    /// Provides functionality for rendering a virtualized, infinite list of items.
+    /// </summary>
+    /// <typeparam name="TItem">The <c>context</c> type for the items being rendered.</typeparam>
+    public sealed class InfiniteScroll<TItem> : VirtualizeBase<TItem>
+    {
+        private readonly ConcurrentQueue<TItem> _loadedItems = new ConcurrentQueue<TItem>();
+
+        private int _fetchingItems;
+
+        protected override int ItemCount => InfiniteScrollFooter == null ? _loadedItems.Count : _loadedItems.Count + 1;
+
+        /// <summary>
+        /// Gets or sets the list item template.
+        /// </summary>
+        [Parameter]
+        public RenderFragment<TItem> Item { get; set; } = default!;
+
+        /// <summary>
+        /// Gets or sets the infinite scroll footer template.
+        /// </summary>
+        [Parameter]
+        public RenderFragment? InfiniteScrollFooter { get; set; }
+
+        /// <summary>
+        /// Gets or sets the function that provides items for the inifinite scroll.
+        /// </summary>
+        [Parameter]
+        public Func<Range, Task<IEnumerable<TItem>>> ItemsProvider { get; set; } = default!;
+
+        /// <summary>
+        /// The batch size in which items are fetched.
+        /// </summary>
+        [Parameter]
+        public int ItemBatchSize { get; set; }
+
+        protected override IEnumerable<RenderFragment> GetItems(int start, int count)
+        {
+            if (start + count >= _loadedItems.Count)
```

Comment:
Is there a risk that we ask the developer to fetch some more, and they do fetch some more, but it's still not enough to fill the vertical height of the component? If so what happens then? I'm wondering if we'll get stuck because the spacer will remain visible and hence no subsequent event will fire to say it's becoming visible.

Reply:
This may have been a bug with my previous implementation, but now this scenario behaves properly. The `MutationObserver` is looking at the "after" spacer, so even if the number of initial items fetched isn't enough to fill the visible space, the `MutationObserver` will catch that the spacer changed height and reset the `IntersectionObserver`, and since the bottom spacer is still intersecting with the container, `OnBottomSpacerVisible` will be invoked, which initiates another data fetch.

And, if the _total_ number of items isn't enough to fill the visible space, `Virtualize` won't get stuck in a loop of constantly requesting more items because the "after" spacer will stop changing height when no more items are loaded.

However, there's currently no good way to proactively fetch more items and alert `Virtualize` that it has more items to render (at least when using `ItemsProvider`). That's probably something worth looking into.